Add async results API for visualizations and refactor dispatch IO

gmarupilla opened the pull request:
Fixes #782

Commit:
Fetch the result structure and results in correct format

Commit:
Plugin optimized API's for visualizations, change env for dispatch URL

Commit:
Use environment files from libs

bilalshaikh42 commented on biosimulations/apps/dispatch-api/src/app/app.controller.ts:
```diff
@@ -138,102 +127,66 @@ export class AppController implements OnApplicationBootstrap {
-          );
+    sedmls.forEach((sedml) => {
+      structure[sedml] = [];
+      const taskFiles = fs.readdirSync(path.join(resultPath, sedml));
```
Could you change these to use an async method? The sync methods will only allow the API to  handle one client at a time

gmarupilla replied:
I can, but by default the async method for file read expects a callback which is fine for a non API use case, we need to return the data as JSON in response which by itself is a synchronous use case in my opinion, do you want me to wrap the async call in a promise and then return that promise? This is the only way that I think which we can make it async.

But internally it would actually by synchronous because NestJS will wait for the promise to get resolved before returning the response, or that won't be the case and I'm missing something here?

bilalshaikh42 replied:
So while nest will resolve the promise, it will be able to handle other user requests. 
Here is a look at how to do this using promised. Essentially just using the await keyword will have node do all the work for you . 
https://nodejs.org/api/fs.html#fs_promise_example

gmarupilla replied:
Done

bilalshaikh42 commented on biosimulations/apps/dispatch-api/src/app/app.controller.ts:
```diff
@@ -138,102 +127,66 @@ export class AppController implements OnApplicationBootstrap {
+      task
+    );
+    const filePath = chart ? `${jsonPath}_chart.json` : `${jsonPath}.json`;
+    const fileContent = JSON.parse(fs.readFileSync(filePath).toString());
```
Same here, we need to switch all of these to async

gmarupilla replied:
Done

Commit:
Make filereading asynchronous

Commit:
Wait for promise resolution instead of returning promise

bilalshaikh42 reviewed (changes requested):
Hi akhil,
You are still using synchrnous methods at various locations (such as line 105 in the upload method). Could you change those?

bilalshaikh42 committed:
Merge branch 'dev' into results_api

Commit:
Make file system methods asynchronous via promise

biosimulations-daemon reviewed (dismissed):
Each iteration of this loop will be blocking due to  the await. A observable/promise followed by a map would be the better approach here

bilalshaikh42 commented:
Lets hold off on creating a library for the utils on this PR. If we create a library, then then nx system starts to diverge and becomes very difficult to integrate. Abstracting the library out can be its own future pr, some code repetition is fine for now

Commit:
Remove dependency of await in iterations

Commit:
Remove  job monitoring, Add temp API for job finish event

gmarupilla commented:
I removed the job monitoring code, added a new API (/api/dispatch-finish) to trigger chart data generation. The chart data won't be generated unless and until the mentioned API is hit.

Commit:
Fix tests for dispatch-api and dispatch-service

Commit:
Add tests for getVisualizationData methods and dispatchFinishEvent method

Commit:
Add tests for PR diff

bilalshaikh42 committed:
Merge branch 'dev' into results_api

bilalshaikh42 committed:
update urls

bilalshaikh42 committed:
accept name and email

bilalshaikh42 committed:
fix method call

bilalshaikh42 committed:
update proxy
